%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional
  \DeclareUnicodeCharacter{"00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{"2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{"2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{"2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{"251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{"2572}{\textbackslash}
 \else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{2572}{\textbackslash}
 \fi
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage[dontkeepoldnames]{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\captionsenglish{\renewcommand{\literalblockcontinuedname}{continued from previous page}}
\addto\captionsenglish{\renewcommand{\literalblockcontinuesname}{continues on next page}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{SpaOTsc}
\date{Jan 18, 2020}
\release{0.2}
\author{}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}



\chapter{SpaOTsc API Reference}
\label{\detokenize{api:module-spaotsc}}\label{\detokenize{api:spaotsc-api-reference}}\label{\detokenize{api::doc}}\label{\detokenize{api:welcome-to-spaotsc-s-documentation}}\index{spaotsc (module)}

\section{The SpaOTsc module}
\label{\detokenize{api:module-spaotsc.SpaOTsc}}\label{\detokenize{api:the-spaotsc-module}}\index{spaotsc.SpaOTsc (module)}\index{choose\_landmarks() (in module spaotsc.SpaOTsc)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.choose_landmarks}}\pysiglinewithargsret{\sphinxcode{spaotsc.SpaOTsc.}\sphinxbfcode{choose\_landmarks}}{\emph{pts}, \emph{n}, \emph{dmat=None}, \emph{method='maxmin'}, \emph{assignment='nearest'}}{}
Choose a set of landmark points from a set of points.

{[}1{]} De Silva, Vin, and Gunnar E. Carlsson. “Topological estimation using
witness complexes.” SPBG 4 (2004): 157-166.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{pts} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} coordinates of points (n\_points, nD) needed if dmat not given

\item {} 
\sphinxstyleliteralstrong{n} (\sphinxstyleliteralemphasis{int}) \textendash{} number of landmark points to select

\item {} 
\sphinxstyleliteralstrong{dmat} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} the distance matrix for the points (n\_points, n\_points)

\end{itemize}

\item[{Returns}] \leavevmode
the indices of selected points and an assignment matrix to assign original points to landmark points (n\_landmarks, n\_points)

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_mcc() (in module spaotsc.SpaOTsc)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.compute_mcc}}\pysiglinewithargsret{\sphinxcode{spaotsc.SpaOTsc.}\sphinxbfcode{compute\_mcc}}{\emph{true\_labels}, \emph{pred\_labels}}{}
Compute matthew’s correlation coefficient.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{true\_labels} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} 1D integer array

\item {} 
\sphinxstyleliteralstrong{pred\_labels} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} 1D integer array

\end{itemize}

\item[{Returns}] \leavevmode
mcc

\item[{Return type}] \leavevmode
float

\end{description}\end{quote}

\end{fulllineitems}

\index{knn\_graph() (in module spaotsc.SpaOTsc)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.knn_graph}}\pysiglinewithargsret{\sphinxcode{spaotsc.SpaOTsc.}\sphinxbfcode{knn\_graph}}{\emph{D}, \emph{k}}{}
Construct a k-nearest-neighbor graph as igraph object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{D} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} a distance matrix for constructing the knn graph

\item {} 
\sphinxstyleliteralstrong{k} (\sphinxstyleliteralemphasis{int}) \textendash{} number of nearest neighbors

\end{itemize}

\item[{Returns}] \leavevmode
a knn graph object

\item[{Return type}] \leavevmode
class:\sphinxtitleref{igraph.Graph}

\end{description}\end{quote}

\end{fulllineitems}

\index{knn\_graph\_nx() (in module spaotsc.SpaOTsc)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.knn_graph_nx}}\pysiglinewithargsret{\sphinxcode{spaotsc.SpaOTsc.}\sphinxbfcode{knn\_graph\_nx}}{\emph{D}, \emph{k}}{}
Construct a k-nearest-neighbor graph as networkx object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{D} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} a distance matrix for constructing the knn graph

\item {} 
\sphinxstyleliteralstrong{k} (\sphinxstyleliteralemphasis{int}) \textendash{} number of nearest neighbors

\end{itemize}

\item[{Returns}] \leavevmode
a knn graph object and a list of edges

\item[{Return type}] \leavevmode
class:\sphinxtitleref{networkx.Graph}, list of tuples

\end{description}\end{quote}

\end{fulllineitems}

\index{phi\_exp() (in module spaotsc.SpaOTsc)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.phi_exp}}\pysiglinewithargsret{\sphinxcode{spaotsc.SpaOTsc.}\sphinxbfcode{phi\_exp}}{\emph{x}, \emph{eta}, \emph{nu}, \emph{p}}{}
The exponential weight kernel. Computes exp(-(x/eta)\textasciicircum{}(p*nu)).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{x} (float or class:\sphinxtitleref{numpy.ndarray}) \textendash{} the input value

\item {} 
\sphinxstyleliteralstrong{eta} (\sphinxstyleliteralemphasis{float}) \textendash{} the cutoff for this soft thresholding kernel

\item {} 
\sphinxstyleliteralstrong{nu} (\sphinxstyleliteralemphasis{int}) \textendash{} a possitive integer for the power term, a bigger nu gives sharper threshold boundary

\item {} 
\sphinxstyleliteralstrong{p} (\sphinxstyleliteralemphasis{int}) \textendash{} p=1: emphasize elements lower than cutoff; p=-1: emphasize elements higher than cutoff

\end{itemize}

\item[{Returns}] \leavevmode
the kernel output with same shape of x

\item[{Return type}] \leavevmode
same as x

\end{description}\end{quote}

\end{fulllineitems}

\index{sci() (in module spaotsc.SpaOTsc)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.sci}}\pysiglinewithargsret{\sphinxcode{spaotsc.SpaOTsc.}\sphinxbfcode{sci}}{\emph{x}, \emph{y}, \emph{W}, \emph{scale=False}}{}
Computes the spatial correlation index in Eq. (9) of {[}1{]}.

{[}1{]} Chen, Yanguang. “A new methodology of spatial cross-correlation
analysis.” PloS one 10.5 (2015): e0126158.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{x} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} the variable’s values at the spatial locations

\item {} 
\sphinxstyleliteralstrong{y} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} the other variable’s values at the spatial locations

\item {} 
\sphinxstyleliteralstrong{W} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} weight matrix (symmetric) among the locations with W{[}i,i{]} = 0

\item {} 
\sphinxstyleliteralstrong{scale} (\sphinxstyleliteralemphasis{boolean}) \textendash{} whether to scale the inputs s.t. (1) sum\_\{i,j\}W\_\{ij\} = 1 and (2) x = (x-mu(x))/sigma(x)

\end{itemize}

\item[{Returns}] \leavevmode
a global spatial cross correlation index

\item[{Return type}] \leavevmode
float

\end{description}\end{quote}

\end{fulllineitems}

\index{spatial\_sc (class in spaotsc.SpaOTsc)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc}}\pysiglinewithargsret{\sphinxbfcode{class }\sphinxcode{spaotsc.SpaOTsc.}\sphinxbfcode{spatial\_sc}}{\emph{sc\_data=None}, \emph{is\_data=None}, \emph{sc\_data\_bin=None}, \emph{is\_data\_bin=None}, \emph{is\_pos=None}, \emph{is\_dmat=None}, \emph{sc\_dmat=None}}{}
An object for connecting and analysis of spatial data and single-cell transcriptomics data.

A minimal example usage:
Assume we have (1) a pandas DataFrame for single-cell data \sphinxcode{df\_sc} with rows being cells and columns being genes
(2) a numpy array for distance matrix among spatial locations \sphinxcode{is\_dmat}
(3) a numpy array for dissimilarity between single-cell data and spatial data \sphinxcode{cost\_matrix}
(4) a numpy array for dissimilarity matrix within single-cell data \sphinxcode{sc\_dmat}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{import} \PYG{n+nn}{spaotsc}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{spsc} \PYG{o}{=} \PYG{n}{spaotsc}\PYG{o}{.}\PYG{n}{SpaOTsc}\PYG{o}{.}\PYG{n}{spatial\PYGZus{}sc}\PYG{p}{(}\PYG{n}{sc\PYGZus{}data}\PYG{o}{=}\PYG{n}{df\PYGZus{}sc}\PYG{p}{,} \PYG{n}{is\PYGZus{}dmat}\PYG{o}{=}\PYG{n}{is\PYGZus{}dmat}\PYG{p}{,} \PYG{n}{sc\PYGZus{}dmat}\PYG{o}{=}\PYG{n}{sc\PYGZus{}dmat}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{spsc}\PYG{o}{.}\PYG{n}{transport\PYGZus{}plan}\PYG{p}{(}\PYG{n}{cost\PYGZus{}matrix}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{spsc}\PYG{o}{.}\PYG{n}{cell\PYGZus{}cell\PYGZus{}distance}\PYG{p}{(}\PYG{n}{use\PYGZus{}landmark}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{spsc}\PYG{o}{.}\PYG{n}{clustering}\PYG{p}{(}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{spsc}\PYG{o}{.}\PYG{n}{spatial\PYGZus{}signaling\PYGZus{}ot}\PYG{p}{(}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Wnt5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fz}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}\PYG{n}{DSgenes\PYGZus{}up}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{CycD}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}\PYG{n}{DSgenes\PYGZus{}down}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{dpp}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{signal\PYGZus{}strengths}\PYG{p}{,}\PYG{n}{\PYGZus{}}\PYG{o}{=}\PYG{n}{spsc}\PYG{o}{.}\PYG{n}{infer\PYGZus{}signal\PYGZus{}range\PYGZus{}ml}\PYG{p}{(}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Wnt5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fz}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{CycD}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{dpp}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{effect\PYGZus{}ranges}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{50}\PYG{p}{,}\PYG{l+m+mi}{100}\PYG{p}{]}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{intercellular\PYGZus{}grn}\PYG{o}{=}\PYG{n}{spsc}\PYG{o}{.}\PYG{n}{spatial\PYGZus{}grn\PYGZus{}range}\PYG{p}{(}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Wnt5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fz}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{CycD}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{dpp}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\end{sphinxVerbatim}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{sc\_data} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} single-cell data of size (n\_cells, n\_genes)

\item {} 
\sphinxstyleliteralstrong{is\_data} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} spatial data of size (n\_locations, n\_genes)

\item {} 
\sphinxstyleliteralstrong{sc\_data\_bin} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} binarized single-cell data

\item {} 
\sphinxstyleliteralstrong{is\_data\_bin} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} binarized spatial data

\item {} 
\sphinxstyleliteralstrong{is\_pos} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} coordinates of spatial locations (n\_locations, n\_dimensions)

\item {} 
\sphinxstyleliteralstrong{is\_dmat} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} distance matrix for spatial locations (n\_locations, n\_locations)

\item {} 
\sphinxstyleliteralstrong{sc\_dmat} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} dissimilarity matrix for single-cell data (n\_cells, n\_cells)

\end{itemize}

\end{description}\end{quote}

List of instance attributes:
\begin{quote}\begin{description}
\item[{Variables}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{sc\_data} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} single-cell data of size (n\_cells, n\_genes) \sphinxcode{\_\_init\_\_}

\item {} 
\sphinxstyleliteralstrong{is\_data} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} spatial data of size (n\_locations, n\_genes) \sphinxcode{\_\_init\_\_}

\item {} 
\sphinxstyleliteralstrong{sc\_data\_bin} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} binarized single-cell data \sphinxcode{\_\_init\_\_}

\item {} 
\sphinxstyleliteralstrong{is\_data\_bin} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} binarized spatial data \sphinxcode{\_\_init\_\_}

\item {} 
\sphinxstyleliteralstrong{is\_pos} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} coordinates of spatial locations (n\_locations, n\_dimensions) \sphinxcode{\_\_init\_\_}

\item {} 
\sphinxstyleliteralstrong{is\_dmat} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} distance matrix for spatial locations (n\_locations, n\_locations) \sphinxcode{\_\_init\_\_}

\item {} 
\sphinxstyleliteralstrong{sc\_dmat} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} dissimilarity matrix for single-cell data (n\_cells, n\_cells) \sphinxcode{\_\_init\_\_}

\item {} 
\sphinxstyleliteralstrong{gamma\_mapping} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} the mapping matrix between single-cell data and spatial data (n\_cells, n\_locations) \sphinxcode{transport\_plan}

\item {} 
\sphinxstyleliteralstrong{sc\_dmat\_spatial} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} the spatial cell-cell distance for single-cell data (n\_cells, n\_cells) \sphinxcode{cell\_cell\_distance}

\item {} 
\sphinxstyleliteralstrong{clustering\_ncluster\_org} (\sphinxstyleliteralemphasis{int}) \textendash{} number of clusters in original clustering of single-cell data \sphinxcode{clustering}

\item {} 
\sphinxstyleliteralstrong{clustering\_nsubcluster} (\sphinxstyleliteralemphasis{list of int}) \textendash{} number of cell spatial subclusters within each original cluster \sphinxcode{clustering}

\item {} 
\sphinxstyleliteralstrong{clustering\_partition\_org} (\sphinxstyleliteralemphasis{list of numpy integer arrays}) \textendash{} the cell indices for each original cluster \sphinxcode{clustering}

\item {} 
\sphinxstyleliteralstrong{clustering\_partition\_inds} (\sphinxstyleliteralemphasis{dictionary}) \textendash{} the cell indices for the cell spatial subclusters, e.g. the key (0,1) returns the cell indices for the second spatial subcluster within the first original cell cluster. \sphinxcode{clustering}

\item {} 
\sphinxstyleliteralstrong{gene\_cor\_scc} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} the intracellular spearmanr correlation between genes \sphinxcode{nonspatial\_correlation}

\item {} 
\sphinxstyleliteralstrong{gene\_cor\_is} (class:\sphinxtitleref{pandas.DataFrame}) \textendash{} the intercellular spatial correlation between genes \sphinxcode{spatial\_correlation}

\item {} 
\sphinxstyleliteralstrong{g\_bin\_edges} (\sphinxstyleliteralemphasis{dictionary}) \textendash{} the bin edges for the discretization of gene expressions with gene name string as dictionary key \sphinxcode{discretize\_expression}

\end{itemize}

\end{description}\end{quote}
\index{cell\_cell\_distance() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.cell_cell_distance}}\pysiglinewithargsret{\sphinxbfcode{cell\_cell\_distance}}{\emph{epsilon=0.01}, \emph{rho=inf}, \emph{scaling=True}, \emph{sc\_dmat\_spatial=None}, \emph{use\_landmark=False}, \emph{n\_landmark=100}}{}
Compute spatial distance between single cells using optimal transport.

Generates: \sphinxtitleref{self.sc\_dmat\_spatial}: (n\_cell, n\_cell) \sphinxtitleref{numpy.ndarray}

Requires: \sphinxtitleref{self.gamma\_mapping}, \sphinxtitleref{self.is\_dmat}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{epsilon} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 0.01}) \textendash{} weight for entropy regularization term

\item {} 
\sphinxstyleliteralstrong{rho} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to inf}) \textendash{} weight for KL divergence penalizing unbalanced transport

\item {} 
\sphinxstyleliteralstrong{scaling} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to True}) \textendash{} whether to scale the cost\_matrix (is\_dmat) to avoid numerical overflow

\item {} 
\sphinxstyleliteralstrong{sc\_dmat\_spatial} (class:\sphinxtitleref{numpy.ndarray}, optional) \textendash{} the spatial distance matrix for single cells (n\_cells, n\_cells). If given, simply set the distance matrix without computing.

\item {} 
\sphinxstyleliteralstrong{use\_landmark} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether to use landmark points for computing transport distance.

\item {} 
\sphinxstyleliteralstrong{n\_landmark} (\sphinxstyleliteralemphasis{int}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 100}) \textendash{} number of landmark points to use if use\_landmark

\end{itemize}

\item[{Returns}] \leavevmode
(spatial) cell-cell distance matrix (n\_cells, n\_cells)

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{clustering() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.clustering}}\pysiglinewithargsret{\sphinxbfcode{clustering}}{\emph{genes=None}, \emph{pca\_n\_components=None}, \emph{res\_sc=0.5}, \emph{res\_is=0.3}, \emph{min\_n=3}}{}
Clustering and spatial subclustering.

Generates:

\sphinxtitleref{self.clustering\_nsubcluster}: list of int, numbers of subclusters in each cluster obtained in regular clustering of single-cell data

\sphinxtitleref{self.clustering\_partition\_inds}: list of cell index arrays for clusters

\sphinxtitleref{self.clustering\_partition\_org}: a dictionary for cell index arrays of spatial subclusters. The key (1,0) gives the first subcluster for the second cluster.

Requires:

\sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.sc\_data}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{genes} (\sphinxstyleliteralemphasis{list}) \textendash{} genes to use when clustering single-cell data. All genes in self.sc\_data are used if not specified.

\item {} 
\sphinxstyleliteralstrong{pca\_n\_components} (\sphinxstyleliteralemphasis{int}) \textendash{} number of pca components when clustering single-cell data

\item {} 
\sphinxstyleliteralstrong{res\_sc} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 0.5}) \textendash{} resolution parameter in louvain clustering for single-cell data

\item {} 
\sphinxstyleliteralstrong{res\_is} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 0.3}) \textendash{} resolution parameter in louvain clustering for spatial subclustering of single-cel data

\item {} 
\sphinxstyleliteralstrong{min\_n} (\sphinxstyleliteralemphasis{int}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 3}) \textendash{} minimum number of members to be considered a cluster

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{discretize\_expression() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.discretize_expression}}\pysiglinewithargsret{\sphinxbfcode{discretize\_expression}}{\emph{genes=None}, \emph{p0=1e-15}}{}
Discretize gene expression using Bayesian blocks.

Generate: \sphinxtitleref{self.g\_bin\_edges}: a dictionary of block edges with gene names as keys

Requires: \sphinxtitleref{self.sc\_data}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{p0} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 1E-15}) \textendash{} the p0 score in Bayesian blocks. A smaller p0 has lower tolerance of false rate, i.e. resulting in fewer blocks.

\end{description}\end{quote}

\end{fulllineitems}

\index{gene\_clustering() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.gene_clustering}}\pysiglinewithargsret{\sphinxbfcode{gene\_clustering}}{\emph{gene\_dmat}, \emph{res=3}, \emph{k=5}, \emph{rng\_seed=48823}}{}
Cluster the genes based on their spatial pattern difference.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{gene\_dmat} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} the distance matrix  for genes (n\_gene, n\_gene)

\item {} 
\sphinxstyleliteralstrong{res} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 3}) \textendash{} resolution parameter used by louvain clustering, higher res gives more clusters

\item {} 
\sphinxstyleliteralstrong{k} (\sphinxstyleliteralemphasis{int}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 5}) \textendash{} the k for knn graph fed to louvain algorithm

\item {} 
\sphinxstyleliteralstrong{rng\_seed} (\sphinxstyleliteralemphasis{int}) \textendash{} random seed for louvain algorithm to get consistent results

\end{itemize}

\item[{Returns}] \leavevmode
a list of index vectors for the clusters

\item[{Return type}] \leavevmode
list of list of int

\end{description}\end{quote}

\end{fulllineitems}

\index{gene\_gene\_distance() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.gene_gene_distance}}\pysiglinewithargsret{\sphinxbfcode{gene\_gene\_distance}}{\emph{genes=None}, \emph{epsilon=0.01}, \emph{rho=inf}, \emph{scaling=True}, \emph{sc\_dmat\_spatial=None}, \emph{use\_landmark=False}, \emph{n\_landmark=100}}{}
Compute Wasserstein distance between gene expressions in scRNA-seq data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{genes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} the gene names to compute distance

\item {} 
\sphinxstyleliteralstrong{epsilon} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 0.01}) \textendash{} weight for entropy regularization term

\item {} 
\sphinxstyleliteralstrong{rho} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to inf}) \textendash{} weight for KL divergence penalizing unbalanced transport

\item {} 
\sphinxstyleliteralstrong{scaling} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to True}) \textendash{} whether to scale the cost matrix

\item {} 
\sphinxstyleliteralstrong{sc\_dmat\_spatial} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} spatial distance matrix over the single cells

\item {} 
\sphinxstyleliteralstrong{use\_landmark} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether to use landmark points to accelarate computation

\item {} 
\sphinxstyleliteralstrong{n\_landmark} (\sphinxstyleliteralemphasis{int}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 100}) \textendash{} number of landmark genes to use

\end{itemize}

\item[{Returns}] \leavevmode
gene-gene distance matrix

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{gene\_pair\_ml\_effect\_range() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.gene_pair_ml_effect_range}}\pysiglinewithargsret{\sphinxbfcode{gene\_pair\_ml\_effect\_range}}{\emph{gene\_1}, \emph{gene\_2}, \emph{background\_genes=None}, \emph{cor\_cut=None}, \emph{n\_top\_g=None}, \emph{effect\_ranges=None}, \emph{method='Importance'}}{}
Deriving scores for intercellular gene regulation (how much effect does gene\_1 in neiborhood have on gene\_2) using random forest.

Requires: \sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.sc\_data}, \sphinxtitleref{self.gene\_cor\_scc}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{gene\_1} (\sphinxstyleliteralemphasis{str}) \textendash{} the name of source gene whose expression in the neighborhood will be examined

\item {} 
\sphinxstyleliteralstrong{gene\_2} (\sphinxstyleliteralemphasis{str}) \textendash{} the name of target gene whose cellular expression will be used

\item {} 
\sphinxstyleliteralstrong{background\_genes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} a name list for gene that are correlated to gene\_2

\item {} 
\sphinxstyleliteralstrong{cor\_cut} (\sphinxstyleliteralemphasis{float}) \textendash{} the cut\_off choosing background genes. used when background\_genes is not specified

\item {} 
\sphinxstyleliteralstrong{n\_top\_g} (\sphinxstyleliteralemphasis{int}) \textendash{} the number of genes with highest correlation to gene\_2 to be used as background\_genes. used when both background\_genes and cor\_cut are not specified

\item {} 
\sphinxstyleliteralstrong{effect\_ranges} (\sphinxstyleliteralemphasis{list of float}) \textendash{} list of spatial distances to consider

\item {} 
\sphinxstyleliteralstrong{method} (\sphinxstyleliteralemphasis{str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 'Importance'}) \textendash{} ‘Importance’: interpret the feature importance as regulation strength; ‘Prediction’: interpret prediction accuracy in cross-validation as regulation strength.

\end{itemize}

\item[{Returns}] \leavevmode
a (n\_distance, 2) array with the first row recording the spatial distances examined and the second row being the effect strength

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{gene\_pair\_pid\_effect\_range() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.gene_pair_pid_effect_range}}\pysiglinewithargsret{\sphinxbfcode{gene\_pair\_pid\_effect\_range}}{\emph{gene\_1}, \emph{gene\_2}, \emph{background\_genes=None}, \emph{cor\_cut=None}, \emph{n\_top\_g=None}, \emph{effect\_ranges=None}, \emph{p0=1e-15}, \emph{cell\_id=None}, \emph{output\_individual=False}}{}~\begin{description}
\item[{The unique information provided by G1\_nb (within various ranges) to}] \leavevmode
G2 considering background genes Gi

\end{description}

Requires: \sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.sc\_data}, \sphinxtitleref{self.gene\_cor\_scc}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{gene\_1} (\sphinxstyleliteralemphasis{str}) \textendash{} the name of source gene whose expression in the neighborhood will be examined

\item {} 
\sphinxstyleliteralstrong{gene\_2} (\sphinxstyleliteralemphasis{str}) \textendash{} the name of target gene whose cellular expression will be used

\item {} 
\sphinxstyleliteralstrong{background\_genes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} a name list for gene that are correlated to gene\_2

\item {} 
\sphinxstyleliteralstrong{cor\_cut} (\sphinxstyleliteralemphasis{float}) \textendash{} the cut\_off choosing background genes. used when background\_genes is not specified

\item {} 
\sphinxstyleliteralstrong{n\_top\_g} (\sphinxstyleliteralemphasis{int}) \textendash{} the number of genes with highest correlation to gene\_2 to be used as background\_genes. used when both background\_genes and cor\_cut are not specified

\item {} 
\sphinxstyleliteralstrong{effect\_ranges} (\sphinxstyleliteralemphasis{list of float}) \textendash{} list of spatial distances to consider

\item {} 
\sphinxstyleliteralstrong{p0} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 1E-15}) \textendash{} the p0 score in Bayesian blocks. A smaller p0 has lower tolerance of false rate, i.e. resulting in fewer blocks.

\item {} 
\sphinxstyleliteralstrong{output\_individual} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} where to output the information computed with each background gene

\end{itemize}

\item[{Returns}] \leavevmode
a (n\_distance, 2) array with the first row recording the spatial distances examined and the second row being the effect strength

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{infer\_signal\_range\_ml() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.infer_signal_range_ml}}\pysiglinewithargsret{\sphinxbfcode{infer\_signal\_range\_ml}}{\emph{Lgenes}, \emph{Rgenes}, \emph{Dgenes}, \emph{n\_top\_g=50}, \emph{effect\_ranges=None}, \emph{method='Importance'}, \emph{custom\_dmat=None}}{}
Determine spatial distance for given signaling using random forest.

Requires: \sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.sc\_data}, \sphinxtitleref{self.gene\_cor\_scc}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{Lgenes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of ligand genes

\item {} 
\sphinxstyleliteralstrong{Rgenes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of receptor genes

\item {} 
\sphinxstyleliteralstrong{Dgenes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of downstream genes

\item {} 
\sphinxstyleliteralstrong{n\_top\_g} (\sphinxstyleliteralemphasis{int}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 50}) \textendash{} number of background genes to use when building predictive model.

\item {} 
\sphinxstyleliteralstrong{effect\_ranges} (\sphinxstyleliteralemphasis{list of float}) \textendash{} the spatial distances to examine

\item {} 
\sphinxstyleliteralstrong{method} (\sphinxstyleliteralemphasis{str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 'Importance'}) \textendash{} the way of interpreting likelihood for each spatial distance

\item {} 
\sphinxstyleliteralstrong{custom\_dmat} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} a cell-cell distance matrix given by user. \sphinxtitleref{self.sc\_dmat\_spatial} is used if not given.

\end{itemize}

\item[{Returns}] \leavevmode
(n\_distance, 2) array for spatial distances (first row) and effect strengths (second row); and a (n\_distance, n\_DSgenes) array for the effect strength of each downstream genes.

\item[{Return type}] \leavevmode
two class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{nonspatial\_correlation() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.nonspatial_correlation}}\pysiglinewithargsret{\sphinxbfcode{nonspatial\_correlation}}{\emph{genes=None}}{}
Compute gene-gene correlation matrix for pre-screening of genes.

Generates: \sphinxtitleref{self.gene\_cor\_scc}

Requires: self.sc\_data{}`, \sphinxtitleref{self.sc\_genes}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{genes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} list of gene names. If not specified, all genes in self.sc\_data are used.

\end{description}\end{quote}

\end{fulllineitems}

\index{rank\_marker\_genes() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.rank_marker_genes}}\pysiglinewithargsret{\sphinxbfcode{rank\_marker\_genes}}{\emph{cid}, \emph{genes=None}, \emph{method='ranksum'}, \emph{return\_scores=False}}{}
Rank genes to identify markers for cell clusters.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{cid} (class:\sphinxtitleref{numpy.1darray}) \textendash{} cell indices for the cluster

\item {} 
\sphinxstyleliteralstrong{genes} (\sphinxstyleliteralemphasis{list}) \textendash{} candidate genes to examine. If not specified, all genes are used.

\item {} 
\sphinxstyleliteralstrong{method} (\sphinxstyleliteralemphasis{str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 'ranksum'}) \textendash{} method to use. 1. ‘roc’, using auc-roc score to rank; 2. ‘ranksum’, using ranksum statistics.

\item {} 
\sphinxstyleliteralstrong{return\_scores} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether to return scores instead of sorted gene indices

\end{itemize}

\item[{Returns}] \leavevmode
sorted gene indices (if return\_scores==False) or gene scores (if return\_scores==True)

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.1darray}

\end{description}\end{quote}

\end{fulllineitems}

\index{spatial\_correlation() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.spatial_correlation}}\pysiglinewithargsret{\sphinxbfcode{spatial\_correlation}}{\emph{genes=None}, \emph{effect\_range=None}, \emph{kernel='lorentz'}, \emph{kernel\_nu=10}}{}
Computes spatial correlation between genes for pre-screening.

Generates: \sphinxtitleref{self.gene\_cor\_is} pandas DataFrame

Requires: \sphinxtitleref{self.sc\_data}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{genes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} list of gene to examine

\item {} 
\sphinxstyleliteralstrong{effect\_range} (\sphinxstyleliteralemphasis{float}) \textendash{} spatial distance

\item {} 
\sphinxstyleliteralstrong{kernel} (\sphinxstyleliteralemphasis{str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 'lorentz'}) \textendash{} type of kernels for weight matrix

\item {} 
\sphinxstyleliteralstrong{kernel\_nu} (\sphinxstyleliteralemphasis{int}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 10}) \textendash{} power for weight kernel

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{spatial\_grn\_range() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.spatial_grn_range}}\pysiglinewithargsret{\sphinxbfcode{spatial\_grn\_range}}{\emph{genes}, \emph{effect\_range=None}, \emph{cor\_cut=None}, \emph{n\_top\_edge=None}, \emph{cor\_cut\_bg=None}, \emph{n\_top\_g\_bg=None}, \emph{method='pid'}, \emph{p0=1e-15}, \emph{output\_individual=False}}{}
Generate the spatial map for intercellular gene-gene regulatory information flow.

Requires: \sphinxtitleref{self.sc\_data}, \sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.gene\_cor\_scc}, \sphinxtitleref{self.gene\_cor\_is}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{genes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of genes to be examined

\item {} 
\sphinxstyleliteralstrong{effect\_range} (\sphinxstyleliteralemphasis{float}) \textendash{} spatial distance for analyzing the intercellular processes

\item {} 
\sphinxstyleliteralstrong{cor\_cut} (\sphinxstyleliteralemphasis{float}) \textendash{} the cutoff for spatial correlation between two genes for further examination (used if n\_top\_edge not specified)

\item {} 
\sphinxstyleliteralstrong{n\_top\_edge} (\sphinxstyleliteralemphasis{int}) \textendash{} the number of gene pairs to examine with highest spatial correlation

\item {} 
\sphinxstyleliteralstrong{cor\_cut\_bg} (\sphinxstyleliteralemphasis{float}) \textendash{} the cutoff for intracellular gene correlation to select background genes

\item {} 
\sphinxstyleliteralstrong{n\_top\_g\_bg} (\sphinxstyleliteralemphasis{int}) \textendash{} the number of genes with highest intracellular gene correlation with the target gene to use as background genes (used if cor\_cut\_bg not specified)

\item {} 
\sphinxstyleliteralstrong{p0} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 1E-15}) \textendash{} the p0 value for Bayesian blocks (lower p0 gives fewer number of bins)

\item {} 
\sphinxstyleliteralstrong{output\_individual} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether to output the individual values computed with each background gene

\end{itemize}

\item[{Returns}] \leavevmode
a data frame with rows being source genes and columns being target genes

\item[{Return type}] \leavevmode
class:\sphinxtitleref{pandas.DataFrame}

\end{description}\end{quote}

\end{fulllineitems}

\index{spatial\_signaling\_ot() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.spatial_signaling_ot}}\pysiglinewithargsret{\sphinxbfcode{spatial\_signaling\_ot}}{\emph{Lgenes}, \emph{Rgenes}, \emph{Tgenes={[}{]}}, \emph{Rbgenes={[}{]}}, \emph{DSgenes\_up={[}{]}}, \emph{DSgenes\_down={[}{]}}, \emph{gene\_bandwidth=\{\}}, \emph{effect\_range=None}, \emph{rho=10.0}, \emph{epsilon=0.2}, \emph{kernel\_nu=5}, \emph{use\_kernel\_ligand=False}, \emph{use\_kernel\_receptor=False}, \emph{return\_weight\_only=False}}{}
Generate cell-cell signaling using optimal transport for a list of ligands and a list of receptors.

Requires: \sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.sc\_data}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{Lgenes} \textendash{} name list of the ligand gene

\item {} 
\sphinxstyleliteralstrong{Rgenes} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of receptor genes

\item {} 
\sphinxstyleliteralstrong{Tgenes} (\sphinxstyleliteralemphasis{list of str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{optional}) \textendash{} name list of genes for transporters of ligands

\item {} 
\sphinxstyleliteralstrong{Rbgenes} (\sphinxstyleliteralemphasis{list of str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{optional}) \textendash{} name list of genes for proteins bound to receptor for the receptor to work

\item {} 
\sphinxstyleliteralstrong{DSgenes\_up} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of up regulated genes by the ligand-receptor

\item {} 
\sphinxstyleliteralstrong{DSgenes\_down} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of down regulated genes by the ligand-receptor

\item {} 
\sphinxstyleliteralstrong{gene\_bandwidth} (\sphinxstyleliteralemphasis{dictionary}\sphinxstyleliteralemphasis{ (}\sphinxstyleliteralemphasis{str to scalar}\sphinxstyleliteralemphasis{)}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{all outputs default to 1}) \textendash{} the cutoffs for each gene to be considered expressed

\item {} 
\sphinxstyleliteralstrong{effect\_range} (\sphinxstyleliteralemphasis{float}) \textendash{} spatial distance cutoff for the signaling

\item {} 
\sphinxstyleliteralstrong{epsilon} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 0.2}) \textendash{} weight for entropy regularization term

\item {} 
\sphinxstyleliteralstrong{rho} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to inf}) \textendash{} weight for KL divergence penalizing unbalanced transport

\item {} 
\sphinxstyleliteralstrong{kernel\_nu} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 5}) \textendash{} the power parameter for the exponential kernel, bigger nu means sharper soft cutoff

\item {} 
\sphinxstyleliteralstrong{use\_kernel\_ligand} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether use kernel function to rescale ligand expression

\item {} 
\sphinxstyleliteralstrong{use\_kernel\_receptor} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether use kernel function to rescale receptor expression

\item {} 
\sphinxstyleliteralstrong{return\_weight\_only} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether to only return the weight for source distribution and destination distribution

\end{itemize}

\item[{Returns}] \leavevmode
a scoring matrix for the given signaling genes (cells, cells), (i,j) entry is the score for cell i sending signals to cell j

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{spatial\_signaling\_ot\_singleligand() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.spatial_signaling_ot_singleligand}}\pysiglinewithargsret{\sphinxbfcode{spatial\_signaling\_ot\_singleligand}}{\emph{Lgene}, \emph{Rgene}, \emph{Tgenes=None}, \emph{Rbgene=None}, \emph{DSgenes\_up=None}, \emph{DSgenes\_down=None}, \emph{effect\_range=None}, \emph{rho=10.0}, \emph{epsilon=0.2}}{}
Generate cell-cell signaling using optimal transport for a single ligand.

Requires: \sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.sc\_data}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{Lgene} (\sphinxstyleliteralemphasis{str}) \textendash{} name of the ligand gene

\item {} 
\sphinxstyleliteralstrong{Rgene} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of receptor genes

\item {} 
\sphinxstyleliteralstrong{Tgenes} (\sphinxstyleliteralemphasis{list of str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{optional}) \textendash{} name list of genes for transporters of ligands

\item {} 
\sphinxstyleliteralstrong{Rbgene} (\sphinxstyleliteralemphasis{list of str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{optional}) \textendash{} name list of genes for proteins bound to receptor for the receptor to work

\item {} 
\sphinxstyleliteralstrong{DSgenes\_up} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of up regulated genes by the ligand-receptor

\item {} 
\sphinxstyleliteralstrong{DSgenes\_down} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of down regulated genes by the ligand-receptor

\item {} 
\sphinxstyleliteralstrong{effect\_range} (\sphinxstyleliteralemphasis{float}) \textendash{} spatial distance cutoff for the signaling

\item {} 
\sphinxstyleliteralstrong{epsilon} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 0.2}) \textendash{} weight for entropy regularization term

\item {} 
\sphinxstyleliteralstrong{rho} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to inf}) \textendash{} weight for KL divergence penalizing unbalanced transport

\end{itemize}

\item[{Returns}] \leavevmode
a scoring matrix for the given signaling genes (cells, cells), (i,j) entry is the score for cell i sending signals to cell j

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{spatial\_signaling\_scoring() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.spatial_signaling_scoring}}\pysiglinewithargsret{\sphinxbfcode{spatial\_signaling\_scoring}}{\emph{Lgene}, \emph{Rgene}, \emph{Rbgene=None}, \emph{Tgenes=None}, \emph{DSgenes\_up=None}, \emph{DSgenes\_down=None}, \emph{effect\_range=None}, \emph{kernel='exp'}, \emph{kernel\_nu=5}, \emph{gene\_eta=None}, \emph{penalty\_type='addition'}}{}
Generate cell-cell signaling using predefined scoring function.

Requires: \sphinxtitleref{self.sc\_dmat\_spatial}, \sphinxtitleref{self.sc\_data}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{Lgene} (\sphinxstyleliteralemphasis{str}) \textendash{} name of the ligand gene

\item {} 
\sphinxstyleliteralstrong{Rgene} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of receptor genes

\item {} 
\sphinxstyleliteralstrong{Rbgene} (\sphinxstyleliteralemphasis{list of str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{optional}) \textendash{} name list of genes for proteins bound to receptor for the receptor to work

\item {} 
\sphinxstyleliteralstrong{Tgenes} (\sphinxstyleliteralemphasis{list of str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{optional}) \textendash{} name list of genes for transporters of ligands

\item {} 
\sphinxstyleliteralstrong{DSgenes\_up} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of up regulated genes by the ligand-receptor

\item {} 
\sphinxstyleliteralstrong{DSgenes\_down} (\sphinxstyleliteralemphasis{list of str}) \textendash{} name list of down regulated genes by the ligand-receptor

\item {} 
\sphinxstyleliteralstrong{effect\_range} (\sphinxstyleliteralemphasis{float}) \textendash{} spatial distance cutoff for the signaling

\item {} 
\sphinxstyleliteralstrong{kernel} (\sphinxstyleliteralemphasis{str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 'exp'}) \textendash{} weight kernel to use for soft thresholding

\item {} 
\sphinxstyleliteralstrong{kernel\_nu} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 5}) \textendash{} power for weight kernel, a higher power gives a shaper edge

\item {} 
\sphinxstyleliteralstrong{gene\_eta} (\sphinxstyleliteralemphasis{list of float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 1s}) \textendash{} a list of threshold values for the downstream genes

\item {} 
\sphinxstyleliteralstrong{penalty\_type} (\sphinxstyleliteralemphasis{str}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 'addition'}) \textendash{} how to penalize inconsistency of downstream genes. ‘addition’: relaxed penalty; ‘multiplication’: strict penalty

\end{itemize}

\item[{Returns}] \leavevmode
a scoring matrix for the given signaling genes (cells, cells), (i,j) entry is the score for cell i sending signals to cell j

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{transport\_plan() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.transport_plan}}\pysiglinewithargsret{\sphinxbfcode{transport\_plan}}{\emph{cost\_matrix}, \emph{cor\_matrix=None}, \emph{alpha=0.1}, \emph{epsilon=1.0}, \emph{rho=100.0}, \emph{G\_sc=None}, \emph{G\_is=None}, \emph{scaling=False}}{}
Mapping between single cells and spatial data as transport plan.

Generates: \sphinxtitleref{self.gamma\_mapping}: (n\_cells, n\_locations) \sphinxtitleref{numpy.ndarray}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{cost\_matrix} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} dissimilarity matrix between single-cell data and spatial data (cells, locations)

\item {} 
\sphinxstyleliteralstrong{cor\_matrix} (class:\sphinxtitleref{numpy.ndarray}, optional) \textendash{} similarity matrix between single-cell data and spatial data (cells, locations)

\item {} 
\sphinxstyleliteralstrong{alpha} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{{[}}\sphinxstyleliteralemphasis{0}\sphinxstyleliteralemphasis{,}\sphinxstyleliteralemphasis{1}\sphinxstyleliteralemphasis{{]}}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 0.1}) \textendash{} weight for structured part (Gromov-Wassertein loss term)

\item {} 
\sphinxstyleliteralstrong{epsilon} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 1.0}) \textendash{} weight for entropy regularization term

\item {} 
\sphinxstyleliteralstrong{rho} (\sphinxstyleliteralemphasis{float}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to 100.0}) \textendash{} weight for KL divergence penalizing unbalanced transport

\item {} 
\sphinxstyleliteralstrong{G\_sc} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} dissimilarity matrix within single-cell data (cells, cells)

\item {} 
\sphinxstyleliteralstrong{G\_is} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} distance matrix within spatial data (locations, locations)

\item {} 
\sphinxstyleliteralstrong{scaling} (\sphinxstyleliteralemphasis{boolean}\sphinxstyleliteralemphasis{, }\sphinxstyleliteralemphasis{defaults to False}) \textendash{} whether scale the cost\_matrix to have max=1

\end{itemize}

\item[{Returns}] \leavevmode
a mapping between single-cell data and spatial data (cells, locations)

\item[{Return type}] \leavevmode
class:\sphinxtitleref{numpy.ndarray}

\end{description}\end{quote}

\end{fulllineitems}

\index{visualize\_cells() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.visualize_cells}}\pysiglinewithargsret{\sphinxbfcode{visualize\_cells}}{\emph{type=1}, \emph{method='umap'}, \emph{perplexity=30.0}, \emph{umap\_n\_neighbors=5}, \emph{umap\_min\_dist=0.1}}{}
Visualization of cells.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{type} (\sphinxstyleliteralemphasis{int}) \textendash{} the type of visualization type=1 dimension reduction with spatial distance, label with original clusters;
type=2 dimension reduction with scRNAseq, label with spatial subclusters;
type=3 dimension reduction with spatial distance, label with spatial subclusters;
type=4 dimension reduction with scRNAseq, label with original clusters.

\end{description}\end{quote}

\end{fulllineitems}

\index{visualize\_subclusters() (spaotsc.SpaOTsc.spatial\_sc method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{api:spaotsc.SpaOTsc.spatial_sc.visualize_subclusters}}\pysiglinewithargsret{\sphinxbfcode{visualize\_subclusters}}{\emph{pts=None}, \emph{k=3}, \emph{cut=None}, \emph{vmin=0.005}, \emph{vmax=0.03333333333333333}, \emph{umap\_k=3}, \emph{figsize=(20}, \emph{20)}}{}
Visualize subclusters as a summary and distributions over the original geometry (2D).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{pts} (class:\sphinxtitleref{numpy.ndarray}) \textendash{} the coordinates of original geometry (n\_locations, 2)

\item {} 
\sphinxstyleliteralstrong{k} (\sphinxstyleliteralemphasis{int}) \textendash{} the number nearest neighbors to connect in the subcluster summary plot

\item {} 
\sphinxstyleliteralstrong{vmin} (\sphinxstyleliteralemphasis{float}) \textendash{} the vmin for colormap of the edges in the summary plot

\item {} 
\sphinxstyleliteralstrong{vmax} (\sphinxstyleliteralemphasis{float}) \textendash{} the vmax for colormap of the edges in the summary plot

\item {} 
\sphinxstyleliteralstrong{umap\_k} (\sphinxstyleliteralemphasis{int}) \textendash{} the n\_neighbors parameter in umap dimension reduction

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{s}
\item {\sphinxstyleindexentry{spaotsc}}\sphinxstyleindexpageref{api:\detokenize{module-spaotsc}}
\item {\sphinxstyleindexentry{spaotsc.SpaOTsc}}\sphinxstyleindexpageref{api:\detokenize{module-spaotsc.SpaOTsc}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}